README.md
# PCEAdvance v7.5

This is an NEC PC Engine / TurboGrafx-16 emulator by FluBBa for the Gameboy Advance with support for PC Engine CD-ROM², rescued from the [Web Archive](https://web.archive.org/web/20150430211123/http://www.ndsretro.com/gbadown.html).

It can also use additional RAM in an EZ-Flash III or SuperCard flashcart to emulate Super CD-ROM² (+192KB), and even Arcade CD-ROM² titles (+2240KB).

### Enhancement
In June 2022 I (patters) forked the source code to [create a version](https://github.com/patters-syno/pceadvance/releases/tag/v7.5-ez4) with the additional RAM support working for EZ-Flash IV and EZ-Flash 3in1 flashcarts. Originally, PCEAdvance enabled the EZ-Flash III PSRAM [here in **cart.s**](https://github.com/patters-syno/pceadvance/blob/65c94787246d6f8c3655c55d0106812b51fa41fc/src/cart.s#L542). This is an ASM implementation of the _OpenRamWrite()_ function as seen in the [firmware source code published by EZ-Team](https://github.com/ez-flash/ez3pda/blob/1d16559caf7a94dff6d8fdcf94f572bc09b36ae4/hard.cpp#L39).

I compared various known-good implementations of PSRAM access for EZ-Flash IV and EZ-Flash 3in1 devices. They were written for the Nintendo DS to access its GBA cartridge port (slot 2), but they will be close enough to serve as useful references:
- Rick "Lick" Wong's [RAM Unlocking API](https://forum.gbadev.org/viewtopic.php?f=18&t=13023) - dead download links now, but the latest v1.3 code can be found in **ram.c** in [memtestARM](http://pineight.com/ds/#memtestARM). This library was used by Simon J Hall's incredible DS ports of [Quake](https://web.archive.org/web/20170115031913/http://quake.drunkencoders.com/index_q1.html) and [Quake II](https://web.archive.org/web/20080912065436/http://quake.drunkencoders.com/index_q2.html).
- Dartz150's [EZFlash3in1 access library](https://github.com/Dartz150/EZFlash3in1/blob/eda7a82ef8ded7103f030a5cdf2d0d4d834b735e/dsCard.cpp#L574)
- Martin (NO$GBA) Korth's [cartridge port RAM documentation](https://www.problemkaputt.de/gbatek-ds-cart-expansion-ram.htm)

They all have in common that they use the [_OpenWrite()_ function](https://github.com/ez-flash/ez3pda/blob/1d16559caf7a94dff6d8fdcf94f572bc09b36ae4/hard.cpp#L19) to unlock the RAM, but _OpenRamWrite()_ is not used. So I substituted _OpenWrite()_ in place of _OpenRamWrite()_ in [PCEAdvance's **cart.s**](https://github.com/patters-syno/pceadvance/blob/e964f93a16d97ecd5427c6353c78da1e7db14d09/src/cart.s#L547), which did indeed awaken the EZ-Flash IV PSRAM!

I then added EZ-Flash specific code for [exit back to the cart menu](https://github.com/patters-syno/pceadvance/blob/ez4/src/visoly.s), re-locking the PSRAM just prior, to replace **visoly.s**. I also added [clearing of the 192KB of PSRAM](https://github.com/patters-syno/pceadvance/blob/e964f93a16d97ecd5427c6353c78da1e7db14d09/src/cart.s#L130) before it is used as emulator RAM. EZ-Flash firmware will [not clear this automatically](https://www.dwedit.org/dwedit_board/viewtopic.php?pid=3349#p3349) so it could contain garbage from a previous game after exit back to the menu. This resolved the issue of Super CD-ROM² support only functioning if loaded directly after a power-on.

Unfortunately it is not possible to add support for EZ-Flash Omega, despite the open nature of its [technical documentation](https://github.com/ez-flash/omega-kernel/blob/master/docs/EZ-FLASH%20OMEGA%20DOCUMENT.pdf). As [confirmed by EZ-Team](https://gbatemp.net/threads/ezflash-omega-writing-in-psram.580813/post-9328539), PSRAM is not exposed for access once in game mode.

### Compatibility

It's mostly slow but there are actually games that are enjoyable:
- 1943 Kai (J) - Takes some time before it starts, but runs ok
- Aero Blasters (U) - ok speed. (spr 6)
- After Burner II (J) - Cool game =)
- Alien Crush (J) - Perfect?
- Atomic Robokid Special (J) - Good speed. (spr 0)
- Bomberman 93 - Perfect?
- Darius (Alpha/Plus) - Very good speed
- Gomola Speed - Strange but funny game
- Hani in the Sky (J) - Good speed
- Image Fight (J) - Good speed
- Kyuukyoku Tiger (J) - Good Speed
- Mr. Heli - Screen to wide, otherwise ok
- Neutopia (U) - Seem ok, even save
- Ninja Warriors (J) - Good speed
- Operation Wolf (J) - Good speed
- Override (J) - Super speed if you turn off Timer IRQ
- Pacland (J) - Good speed
- R-Type - Good speed
- Rastan Saga II (J) - Good speed
- Super Star Soldier
- Tatsujin (J) - Good speed
- Tenseiryu Saint Dragon (J) - Good speed

These are just suggestions, please try what ever game you like. A lot of US games don't work because they are encrypted, use PCEToy to decrypt these before you use them. Also remember to click "US rom" in the legacy Win32 builder app if you want them to work. If using the Python 3 builder make sure US ROM filenames contain (U) or (USA). Don't use overdumps as these are evil on PC Engine.

## How to use
When the emulator starts use Up/Down to select game, then use B or A to start the game selected. Press L+R to open the menu, A to choose, B (or L+R again) to cancel.

### Settings:
	HScroll: (Manual) Lets you scroll the screen with the L & R buttons
	Unscaled modes: L & R buttons scroll the screen up and down
	Scaled modes: Press L+SELECT to adjust the background
	Sound:
		Off: no sound
		On: low quality low CPU usage
		On: (Mixer) - better quality more cpu usage
	TimerIRQ:
		Some games use the TimerIRQ to play sounds and music, by disabling the timer
		you can make some games faster
	EWRAM speed:
		This changes the waitstate on EWRAM between 2 and 1, this can probably damage
		your GBA and definitly uses more power, little to no speedgain. Don't use!
	Speed modes: L+START switches between throttled/unthrottled/slomo mode
	Sleep: START+SELECT wakes up from sleep mode (activated from menu or 5/10/30
	       minutes of inactivity)

## Pogoshell
To use as a Pogoshell plugin, first copy *pceadvance.gba* to the plugin folder then rename it to *pce.bin*. To make it work with US roms the name of the rom must not contain (J) or (j). In the same manner Japanese roms should preferably contain (J) or (j), most Japanese roms seem to run on US hardware anyway though.

## Multiplayer link play
Go to the menu and change Controller: to read *Link2P/Link3P/Link4P*, depending on how many Gameboys you will use. Once this is done on all GBAs, leave the menu on all slaves first, then the master, the game will restart and you can begin playing. If the link is lost (cable is pulled out, or a GBA is restarted), link must be re-initiated, this is done by a restart on the master and then selecting the appropriate link and leave the menu. The slaves doesn't have to do anything. Use an original Nintendo cable!

## SRAM
The first 8KB of the GBA SRAM is the PC Engine SRAM. This can be exchanged between other PC Engine emulators, I think you have to change MagicEngine's INI to old format. Use a CD-ROM System ROM to manage your PC Engine SRAMs, press Select to access the SRAM manager. The US version is encrypted, don't forget to decrypt it.

## PC Engine CD-ROM support
Compilations built with the legacy Win32 builder included with PCEAdvance 7.5 cannot have CD-ROM data appended correctly. This builder mistakenly pads the preceding ROM data which breaks CD support. Use the new Python 3 builder instead. You can read the builder's full help text using the ```-h``` option.

To be able to use PC Engine / TurboGrafx16 CD-ROM games you have to have a CD-ROM System ROM in your build. The builder will add this automatically, it defaults to importing the file *bios.bin* but this can be overridden using the ```-b``` option (BIOS). To use CD-ROM support from Pogoshell just make a compilation with only a CD-ROM System ROM and use it as the plugin for ```.iso``` and ```.pce``` files.

Most CD-ROM games have their data stored in CD track 2, and have a very similar sized second copy of that data as the final CD track. All other tracks are usually audio. PCEAdvance cannot play the audio so usually it only needs a game's track 2 data in ```.iso``` format. This can be extracted from a typical ```.bin/.cue``` disc image using a tool such as Isobuster for Windows, or using *bchunk* on macOS or Linux. Add ```.iso``` files using the builder in the same way as you would add a regular ```.pce``` file.

Some games do have multiple data tracks (excluding the last duplicate of track 2), for instance *Macross 2036*, and in this case they will need a ```.tcd``` track index file. Some are included with PCEAdvance, along with details of the specification. If you need to make new ones, the Table of Contents (TOC) LBA values can be taken directly from https://www.necstasy.net and converted to hex. If the Python 3 builder finds a ```.tcd``` file with the same name as the ```.iso``` file it will be added automatically. The track index can also be manually specified using the ```-t``` option.

Owing to the way PCEAdvance organises the CD-ROM data you are limited to a single CD game in each build, but it can co-exist with other ROMs and it can be added in any order in the list using the Python 3 builder.

Note that PSRAM on the EZ-Flash flashcarts is limited to 16MB. Unfortunately PSRAM cannot be addressed if the emulator is run from NOR flash (32MB). This means that both the PCEAdvance compilation and its additional RAM requirement must fit within that 16MB. Oversized compilations can be truncated by the builder using the ```-trim``` option, losing some game data in the process. For this reason you should label your ISO filenames with the required system type: (CD) for CD-ROM², (SCD) for Super CD-ROM², or (ACD) for Arcade CD-ROM². You can determine this by consulting the lists published at https://www.necstasy.net. *Akumajou Dracula X: Chi no Rondo* (20.8MB) is one such title. Although the trimmed game does apparently work, it would not be playable to completion on EZ-Flash.

#### CD-ROM² games tested so far:
- Addams Family (U): Ok, fullscreen images flicker
- Cosmic Fantasy 2 (U): Intro & game ok, can't fit whole game though
- Download 2 (J): Ok
- Exile (U): Crashes if you hit the Ants
- Final Zone II (U): Ok, need to skip intro
- Gain Ground: Too big.
- Golden Axe: Ok, need to skip intro
- HellFire S: Ok, screen too wide though
- Jyuohki (J)/(Altered Beast): Ok
- Macross 2036 (J): Ok
- MineSweeper (J): Ok
- Monster Lair: Ok
- Rainbow Islands (J): Very slow
- Rayxanber II (U): Ok
- Red Alert (J): Ok
- Road Spirits: Ok
- Space Fantasy Zone (J/U): Ok
- Splash Lake (U): Ok
- Spriggan (J): Ok, stops after 3rd level?
- Super Darius: Ok. What is different from the Hucard version? A bigger logo?
- Valis II (U): Ok, need to skip intro
- Valis III (U): Works,I've only got the first data track so the intro is corrupt
- Valis IV (J): Same as Valis III
- Ys Book 1&2 (U): Ok
- Ys 3: Wanderers From Ys (U): Too big too fit on a flashcart
- Zero Wing (J): Ok

#### Super CD-ROM² games tested so far (SuperCard / EZ-Flash builds only):
- Akumajou Dracula X: Chi no Rondo (J): Ok
- Akumajou Dracula X: Chi no Rondo (J) (T+Eng): Ok
- Conan: Intro Ok
- Cotton - Fantastic Night Dream (U): Ok
- Double Dragon 2: Very slow
- Forgotten Worlds (J): Ok
- Gate of Thunder (J): Ok
- Genocide (J): Ok
- Gradius II - Gofer no Yabou (J): Ok
- Image Fight 2 (U): Ok
- Loom (U): Flickering graphics.
- Lords Of Thunder: Ok
- Nexzr: Ok
- Rayxanber III (J): Ok
- Riot Zone: Ok
- R-Type Complete CD (J): Ok
- Shadow of the Beast (U): Ok, some flicker in intro

#### Arcade CD-ROM² games tested so far (SuperCard / EZ-Flash builds only):
None confirmed working with EZ-Flash at least. Most data tracks are way too large. The only realistic contenders are:
- Mad Stalker (5.1MB): Hangs at loading screen
- Ginga Fukei Densetsu Sapphire (15.6MB): Hangs at white screen
- World Heroes 2 (17.7MB): Hangs at black screen

## Credits
Huge thanks to Loopy for the incredible PocketNES, without it this emulator would probably never have been made. Big thanks to Hoe for the ROM-Builder.
Thanks to:
- Zeograd for a lot of help with the debugging
- [Charles MacDonald](http://techno-junk.org) &
- David Shadoff for a lot of the info.


**Fredrik Ahlström**

https://github.com/FluBBaOfWard

https://twitter.com/TheRealFluBBa

Some things to consider regarding this emulation:
PC Engine|Gameboy Advance
:----|:----
64KB of VRAM which can be background and/or sprites|64KB background VRAM and 32KB sprite VRAM
CPU runs at either 1.78MHz (like the NES) or at 7.2MHz <br />all games seem to use the fast mode|CPU runs at 16MHz


history.txt
PCEAdvance revision history
-=-=-=-=-=-=-=-=-=-=-=-=-=-

V7.5_EZ4 - 2022-06-05 (patters)
   Supports using EZ-Flash IV and EZ-Flash 3in1 PSRAM for Super CD-ROM games
   Supports exit to EZ-Flash menu

V7.5 - 2005-10-14 (FluBBa)
   Optimized DMA control (sprite bit) a bit.
   Fixed SUBQ_READ in CD emulation (FinalZone II intro doesn't crash).
   Fixed AD_PLAY status in CD emulation (Valis II intro doesn't crash).
   Fixed Game Over screen in Rayxanber II.
   Separate version for EZ3.

V7.4 - 2005-08-18 (FluBBa)
   Added Black'n White mode to games that support it.
   Fixed Arcade card emulation (on EZ3 cards).

V7.3 - 2005-02-26 (FluBBa)
   Optimized screen rendering some more.
   Better SCD emulation when no extra memory found.
   Fixed manual scroll when screen < 160 pixels high or < 240 pixels wide.
   Fixed graphics when switching between scale modes.
   Major code cleanup.

V7.2 - 2005-01-30 (FluBBa)
   Optimized screen rendering.
   Fixed a transfer instruction to work with SCD RAM.

V7.1 - 2005-01-19 (FluBBa)
   Added Super CD-Rom emulation on EZ3 cards.
   Optimized VDC write.
   Optimized VRAM read/write.

V7.0 - 2004-11-30 (FluBBa)
   Added software sound mixer.
   Added check for Nintendo DS.
   Fixed a bug where emulated game could write to GBA ROM.
   Fixed clock on EZ3 cards.
   Fixed ROL & ROR instructions a bit.
   Fixed ADC & SBC instructions a bit.
   Fixed VRAM to VRAM DMA a bit.
   Fixed screen height in Unscaled(Auto) mode.
   Optimized VRAM read/write.

V6.9 - 2004-08-22 (FluBBa)
   Added support for iso files from Pogoshell.
   Removed sprite setting from menu.

V6.8 - 2004-08-19 (FluBBa)
   Added correct sound pitch on channel 3 & 4.
   Fixed the gamma adjuster a bit (more to do with the source).
   Better VBLankWait.

V6.7 - 2004-07-18 (FluBBa)
   Added swapping of A & B buttons.
   Fixed transfer on EZFA cards when bootmenu used.
   Minor CPU and tilerender optimizations.
   Speedhacks (Thanks to DKL):
    "Tower of Druaga (J)".
    "Volfied (J)".
    "Wonder Momo (J)".
    "World Circuit (J)".
    "Thunderblade (J)".
    "Timeball (U) / Blodia (J)".
    "Kaiser's Quest (J)".

V6.6 - 2004-06-19 (FluBBa) (Special GBAX compo release =)
   Added double buffering for VDCCNT (Chase HQ).
   Added double buffering for the palette (After Burner II).
   Added a menu option for FPS meter.
   Added a menu option for noflicker scaling.
   Fixed a problem with the CLI instruction in the CPU.
   Added more TCD files.
   Speedhacked Cadash (thanks to DKL).

V6.5 - 2004-05-22 (FluBBa)
   Fixed the multi datatrack support in the .tcd files.

V6.4 - 2004-05-15 (FluBBa)
   Fixed "Davis Cup Tennis" and "Lode Runner".
   Fixed the static in the sound.
   Fixed a potential crash (affected encrypted "Gunboat (U)").
   Rewrote parts of the CPU, should be a tiny bit faster.

V6.3 - 2004-04-23 (FluBBa)
   Added basic CD-ROM emulation.
   Fixed the CLI instruction, TV Sports Hockey seem better, OutRun faster?
   Fixed the ADC when D & T bit set (Rayxanber2).
   Fixed "indexed indirect X" addressing to be more accurate.
   Changed D & I location to register instead of memory.

V6.2 - 2004-03-29 (FluBBa)
   Added new settings menu.
   Fixed graphics bug in Pacland.

V6.1 - 2004-03-19 (FluBBa)
   Fixed crashes in Puzznic and Boxyboy (when run in J mode).

V6.0 - 2004-03-07 (FluBBa)
   Fixed trash when exiting to Pogoshell.
   Changed the menu layout and behaviour.
   Compressed the font (save 1kByte =)

V5.9 - 2004-02-27 (FluBBa)
   Added HScroll option in menu, use L & R.
   Added double buffering of background.
   Fixed some sprite updating (sprite option should be redundant).
   Fixed SRAM initialization.
   Fixed more problems with folders under pogo.
   Fixed multibooting from PogoPCE and on XG cards.
   Changed the default values for Gamma and Scaling.
   Changed detection of max screen height:
    "Chase HQ (U)" starts.
    "Air Zonk (U)" scrolling is ok.
   Speedhacked:
    "Dungeon Explorer".
    "Final Lap Twin".
    "Galaga", hopefully doesn't break it. ;)
    "Jyuohki (Altered Beast)".
    "Blazing Lazers/Gunhed" (intro).
    And a lot more...

V5.5 - 2004-01-15 (FluBBa)
   Added speedhacks for A LOT of games.
   Optimized all the fransfer operands.
   Fixed the "one folder under the root" problem in Pogoshell.
   Fixed the TRB & TSB operands. "Parasol Stars" works.

V5.4 - 2004-01-11 (FluBBa)
   Fixed LSR from zeropage,x (Gunhed(J)/Blazing Lazers(U)).
   Fixed multibooting from multibooted GBA.

V5.3 - 2004-01-07 (FluBBa)
   Fixed more with the Timer IRQ.
   Added support for multibooting 128kbyte roms.

V5.2 - 2004-01-06 (FluBBa)
   Made one of the transfers to VRAM HLE (faster).
   Fixed the Timer IRQ.
   Added new item to menu (TimerIRQ: enable/dissable).

V5.1 - 2003-11-14 (FluBBa)
   Fixed the clock routines.

V5.0 - 2003-11-07 (FluBBa)
   Added clock in menu if your card/emu supports it.
   Resets more hardware regs before exiting.
   Workaround for the XG2 bootmenu.

V4.9 - 2003-10-26 (FluBBa)
   Fixed lockup when no game was found.
   Optimized writes to Y-Scroll reg.
   Added a little fade when entering the menu.
   Added Exit to normal PCEAdvance.
   Now saves config also when exiting.
   B-button is not recognised when leaving menu.

V4.8 - 2003-09-23 (FluBBa)
   Fixed some of the flickering/missing sprites, new menu option.
   Fixed reseting of joypad input.

V4.7 - 2003-08-28 (FluBBa)
   Added checking for GameBoy Player, see readme.

V4.6 - 2003-07-20 (FluBBa)
   Fixed the SRAM bug.

V4.5 - 2003-07-17 (FluBBa)
   Simplified the SRAM handling. Use a "CD-ROM System" rom to manage your save.
   Added multiplayer support.

V4.4 - 2003-07-11 (FluBBa)
   Added speedhacks for SF2CE.
   Fixed "Air Zonk/PC Denjin Cyper punkics" & Toy Shop Boys.

V4.3 - 2003-07-07 (FluBBa)

ui.c
#include <stdio.h>
#include <string.h>

#include "gba.h"

//header files?  who needs 'em :P

void cls(int);		//from main.c
void initpalette(void);
void rommenu(void);
void drawtext(int,char*,int);
void setdarknessgs(int dark);
void setbrightnessall(int light);
extern char *textstart;

int SendMBImageToClient(void);	//mbclient.c

//----asm calls------
void resetSIO(u32);			//io.s
void doReset(void);			//io.s
void suspend(void);			//io.s
void waitframe(void);		//io.s
int gettime(void);			//io.s
void spriteinit(char);		//io.s
void debug_(int,int);		//vdc.s
void paletteinit(void);		//vdc.s
void PaletteTxAll(void);	//vdc.s
void endframe(void);		//vdc.s
void Sound_reset_(void);	//sound.s
//-------------------

extern u32 joycfg;			//from io.s
extern u32 timermask;		//from io.s
extern char g_emuflags;		//from cart.s
extern char g_scaling;		//from cart.s
extern char g_scalingx;		//from cart.s
extern char novblankwait;	//from h6280.s
extern u32 sleeptime;		//from h6280.s
extern u32 FPSValue;		//from vdc.s
extern u32 oambufferready;	//from vdc.s
extern char fpsenabled;		//from vdc.s
extern char gammavalue;		//from vdc.s
extern char twitch;			//from vdc.s
extern char flicker;		//from vdc.s
extern u32 soundmode;		//from sound.s	0=off, 1=on(GB), 2=on(DS)

extern char rtc;
extern char pogoshell;
extern char gameboyplayer;
extern char gbaversion;

u8 autoA,autoB;				//0=off, 1=on, 2=R
u8 stime=0;
u8 ewram=0;

void autoAset(void);
void autoBset(void);
void swapAB(void);
void controller(void);
void vblset(void);
void restart(void);
void exit(void);
void multiboot(void);
void scrolll(int f);
void scrollr(void);
void drawui1(void);
void drawui2(void);
void drawui3(void);
void subui(int menunr);
void ui2(void);
void ui3(void);
void drawclock(void);
void sleep(void);
void sleepset(void);
void fpsset(void);
void brightset(void);
void fadetowhite(void);
void ewramset(void);
void display(void);
void flickset(void);
void soundset(void);
void centerset(void);
void timerset(void);

void writeconfig(void);	//sram.c

#define MENU2ITEMS 6			//othermenu items
#define MENU3ITEMS 4			//displaymenu items
#define CARTMENUITEMS 10		//mainmenuitems when running from cart (not multiboot)
#define MULTIBOOTMENUITEMS 9	//"" when running from multiboot
const fptr fnlist1[]={autoBset,autoAset,controller,ui3,ui2,soundset,multiboot,sleep,restart,exit};
const fptr fnlist2[]={vblset,fpsset,sleepset,ewramset,swapAB,timerset};
const fptr fnlist3[]={display,flickset,brightset,centerset};

int selected;//selected menuitem.  used by all menus.
int mainmenuitems;//? or CARTMENUITEMS, depending on whether saving is allowed

u32 oldkey;//init this before using getmenuinput
u32 getmenuinput(int menuitems) {
	u32 keyhit;
	u32 tmp;
	int sel=selected;

	waitframe();		//(polling REG_P1 too fast seems to cause problems)
	tmp=~REG_P1;
	keyhit=(oldkey^tmp)&tmp;
	oldkey=tmp;
	if(keyhit&UP)
		sel=(sel+menuitems-1)%menuitems;
	if(keyhit&DOWN)
		sel=(sel+1)%menuitems;
	if(keyhit&RIGHT) {
		sel+=10;
		if(sel>menuitems-1) sel=menuitems-1;
	}
	if(keyhit&LEFT) {
		sel-=10;
		if(sel<0) sel=0;
	}
	if((oldkey&(L_BTN+R_BTN))!=L_BTN+R_BTN)
		keyhit&=~(L_BTN+R_BTN);
	selected=sel;
	return keyhit;
}

void ui() {
	int key,soundvol,oldsel,tm0cnt,i;
	ewram=((REG_WRWAITCTL & 0x0F000000) == 0x0E000000)?1:0;

	autoA=joycfg&A_BTN?0:1;
	autoA|=joycfg&(A_BTN<<16)?0:2;
	autoB=joycfg&B_BTN?0:1;
	autoB|=joycfg&(B_BTN<<16)?0:2;

	mainmenuitems=((u32)textstart>0x8000000?CARTMENUITEMS:MULTIBOOTMENUITEMS);//running from rom or multiboot?
	FPSValue=0;					//Stop FPS meter

	soundvol=REG_SGCNT0_L;
	REG_SGCNT0_L=0;				//stop sound (GB)
	tm0cnt=REG_TM0CNT;
	REG_TM0CNT=0;				//stop sound (directsound)

	selected=0;
	drawui1();
	initpalette();
	for(i=0;i<8;i++)
	{
		waitframe();
		setdarknessgs(i);		//Darken game screen
		REG_BG2HOFS=224-i*32;	//Move screen right
	}

	oldkey=~REG_P1;			//reset key input
	do {
		drawclock();
		key=getmenuinput(mainmenuitems);
		if(key&(A_BTN)) {
			oldsel=selected;
			fnlist1[selected]();
			selected=oldsel;
		}
		if(key&(A_BTN+UP+DOWN+LEFT+RIGHT))
			drawui1();
	} while(!(key&(B_BTN+R_BTN+L_BTN)));
	writeconfig();			//save any changes
	for(i=1;i<9;i++)
	{
		waitframe();
		setdarknessgs(8-i);	//Lighten screen
		REG_BG2HOFS=i*32;	//Move screen left
	}
	PaletteTxAll();			//restore PCE palette
	oambufferready=1;
	while(key&(B_BTN)) {
		waitframe();		//(polling REG_P1 too fast seems to cause problems)
		key=~REG_P1;
	}
	REG_SGCNT0_L=soundvol;	//resume sound (GB)
	REG_TM0CNT=tm0cnt;		//resume sound (directsound)
	cls(3);
}

void subui(int menunr) {
	int key,oldsel;

	selected=0;
	if(menunr==2)drawui2();
	if(menunr==3)drawui3();
	scrolll(0);
	oldkey=~REG_P1;			//reset key input
	do {
		if(menunr==2)key=getmenuinput(MENU2ITEMS);
		if(menunr==3)key=getmenuinput(MENU3ITEMS);
		if(key&(A_BTN)) {
			oldsel=selected;
			if(menunr==2)fnlist2[selected]();
			if(menunr==3)fnlist3[selected]();
			selected=oldsel;
		}
		if(key&(A_BTN+UP+DOWN+LEFT+RIGHT)) {
			if(menunr==2)drawui2();
			if(menunr==3)drawui3();
		}
	} while(!(key&(B_BTN+R_BTN+L_BTN)));
	scrollr();
	while(key&(B_BTN)) {
		waitframe();		//(polling REG_P1 too fast seems to cause problems)
		key=~REG_P1;
	}
}

void ui2() {
	subui(2);
}
void ui3() {
	subui(3);
}

void text(int row,char *str) {
	drawtext(row+10-mainmenuitems/2,str,selected==row);
}
void text2(int row,char *str) {
	drawtext(35+row+2,str,selected==row);
}


//trying to avoid using sprintf...  (takes up almost 3k!)
void strmerge(char *dst,char *src1,char *src2) {
	if(dst!=src1)
		strcpy(dst,src1);
	strcat(dst,src2);
}

char *const autotxt[]={"OFF","ON","with R"};
char *const vsynctxt[]={"ON","OFF","SLOWMO"};
char *const sleeptxt[]={"5min","10min","30min","OFF"};
char *const brightxt[]={"I","II","III","IIII","IIIII"};
char *const memtxt[]={"Normal","Turbo"};
char *const hostname[]={"Crap","Prot","GBA","GBP","NDS"};
char *const ctrltxt[]={"1P","2P","Link2P","Link3P","Link4P"};
char *const disptxt[]={"UNSCALED","UNSCALED (Auto)","SCALED","SCALED (w/sprites)"};
char *const flicktxt[]={"No Flicker","Flicker"};
char *const soundtxt[]={"OFF","ON","ON(Mixer)"};
char *const centxt[]={"Centered","Manual","Auto"};
char *const cntrtxt[]={"Pc-Engine (J)","TurboGrafx 16 (U)"};
char *const enabtxt[]={"Disabled","Enabled"};
char *const emuname[]={"      PCEAdvance ","         PogoPCE "};
void drawui1() {
	int i=0;
	char str[30];

	cls(1);
	drawtext(18,"Powered by XGFLASH2.com 2005",0);
	if(pogoshell) i=1;
	strmerge(str,emuname[i],"v7.5 on ");
	strmerge(str,str,hostname[gbaversion]);
	drawtext(19,str,0);

	strmerge(str,"B autofire: ",autotxt[autoB]);
	text(0,str);
	strmerge(str,"A autofire: ",autotxt[autoA]);
	text(1,str);
	strmerge(str,"Controller: ",ctrltxt[(joycfg>>29)-2]);
	text(2,str);
	text(3,"Display->");
	text(4,"Other Settings->");
	strmerge(str,"Sound: ",soundtxt[soundmode]);
	text(5,str);
	text(6,"Link Transfer");
	text(7,"Sleep");
	text(8,"Restart");
	if(mainmenuitems!=MULTIBOOTMENUITEMS) {
		text(9,"Exit");
	}
}

void drawui2() {
	char str[30];

	cls(2);
	drawtext(32,"       Other Settings",0);
	strmerge(str,"VSync: ",vsynctxt[novblankwait]);
	text2(0,str);
	strmerge(str,"FPS-Meter: ",autotxt[fpsenabled]);
	text2(1,str);
	strmerge(str,"Autosleep: ",sleeptxt[stime]);
	text2(2,str);
	strmerge(str,"EWRAM speed: ",memtxt[ewram]);
	text2(3,str);
	strmerge(str,"Swap A-B: ",autotxt[(joycfg>>10)&1]);
	text2(4,str);
	strmerge(str,"TimerIRQ: ",enabtxt[timermask]);
	text2(5,str);
	strmerge(str,"Machine: ",cntrtxt[(g_emuflags & 4)>>2]);		//USCOUNTRY=4
	text2(6,str);
}

void drawui3() {
	char str[30];

	cls(2);
	drawtext(32,"      Display Settings",0);
	strmerge(str,"Display: ",disptxt[g_scaling&3]);
	text2(0,str);
	strmerge(str,"Scaling: ",flicktxt[flicker]);
	text2(1,str);
	strmerge(str,"Gamma: ",brightxt[gammavalue]);
	text2(2,str);
	strmerge(str,"HScroll: ",centxt[g_scalingx&1]);
	text2(3,str);
}

void drawclock() {

    char str[30];
    char *s=str+20;
    int timer,mod;

    if(rtc)
    {
	strcpy(str,"                    00:00:00");
	timer=gettime();
	mod=(timer>>4)&3;				//Hours.
	*(s++)=(mod+'0');
	mod=(timer&15);
	*(s++)=(mod+'0');
	s++;
	mod=(timer>>12)&15;				//Minutes.
	*(s++)=(mod+'0');
	mod=(timer>>8)&15;
	*(s++)=(mod+'0');
	s++;
	mod=(timer>>20)&15;				//Seconds.
	*(s++)=(mod+'0');
	mod=(timer>>16)&15;
	*(s++)=(mod+'0');

	drawtext(0,str,0);
    }
}

void autoAset() {
	autoA++;
	joycfg|=A_BTN+(A_BTN<<16);
	if(autoA==1)
		joycfg&=~A_BTN;
	else if(autoA==2)
		joycfg&=~(A_BTN<<16);
	else
		autoA=0;
}

void autoBset() {
	autoB++;
	joycfg|=B_BTN+(B_BTN<<16);
	if(autoB==1)
		joycfg&=~B_BTN;
	else if(autoB==2)
		joycfg&=~(B_BTN<<16);
	else
		autoB=0;
}

void controller() {					//see io.s: refreshPCEjoypads
	u32 i=joycfg+0x20000000;
	if(i>=0xe0000000)
		i-=0xa0000000;
	resetSIO(i);					//reset link state
}

void sleepset() {
	stime++;
	if(stime==1)
		sleeptime=60*60*10;			// 10min
	else if(stime==2)
		sleeptime=60*60*30;			// 30min
	else if(stime==3)
		sleeptime=0x7F000000;		// 360days...
	else if(stime>=4){
		sleeptime=60*60*5;			// 5min
		stime=0;
	}
}

void vblset() {
	novblankwait++;
	if(novblankwait>=3)
		novblankwait=0;
}

void fpsset() {
	fpsenabled = (fpsenabled^1)&1;
}

void brightset() {
	gammavalue++;
	if (gammavalue>4) gammavalue=0;
	paletteinit();
	PaletteTxAll();					//make new palette visible
	initpalette();
}

void multiboot() {
	int i;
	cls(1);
	drawtext(9,"          Sending...",0);
	i=SendMBImageToClient();
	if(i) {
		if(i<3)
			drawtext(9,"         Link error.",0);
		else
			drawtext(9,"  Game is too big to send.",0);
		if(i==2) drawtext(10,"       (Check cable?)",0);
		for(i=0;i<90;i++)			//wait a while
			waitframe();
	}
}

void restart() {
	writeconfig();					//save any changes
	scrolll(1);
	__asm {mov r0,#0x3007f00}		//stack reset
	__asm {mov sp,r0}
	rommenu();
}
void exit() {
	writeconfig();					//save any changes
	fadetowhite();
	REG_DISPCNT=FORCE_BLANK;		//screen OFF
	REG_BG0HOFS=0;
	REG_BG0VOFS=0;
	REG_BLDCNT=0;					//no blending
	doReset();
}

void sleep() {
	fadetowhite();
	suspend();
	setdarknessgs(7);				//restore screen
	while((~REG_P1)&0x3ff) {
		waitframe();				//(polling REG_P1 too fast seems to cause problems)
	}
}
void fadetowhite() {
	int i;
	for(i=7;i>=0;i--) {
		setdarknessgs(i);			//go from dark to normal
		waitframe();
	}
	for(i=0;i<17;i++) {				//fade to white
		setbrightnessall(i);		//go from normal to white
		waitframe();
	}
}

void scrolll(int f) {
	int i;
	for(i=0;i<9;i++)
	{
		if(f) setdarknessgs(8+i);	//Darken screen
		REG_BG2HOFS=i*32;			//Move screen left
		waitframe();
	}
}
void scrollr() {
	int i;
	for(i=8;i>=0;i--)
	{
		waitframe();
		REG_BG2HOFS=i*32;			//Move screen right
	}
	cls(2);							//Clear BG2
}

void ewramset() {
	ewram^=1;
	if(ewram==1){
		REG_WRWAITCTL = (REG_WRWAITCTL & ~0x0F000000) | 0x0E000000;		//1 waitstate, overclocked
	}else{
		REG_WRWAITCTL = (REG_WRWAITCTL & ~0x0F000000) | 0x0D000000;		//2 waitstates, normal
	}
}

void swapAB() {
	joycfg^=0x400;
}

void display() {
	char sc;
	g_scaling=sc=(g_scaling+1)&3;
	spriteinit(sc);
	endframe();
	initpalette();
}

void flickset() {
	flicker++;
	if(flicker > 1){
		flicker=0;
		twitch=0;
	}
}

//void setcountry() {
//	g_emuflags ^= 4;				// USCOUNTRY
//	if(g_emuflags & 4)				// USCOUNTRY
//		joyptr=&JOYP_Ru;
//	else
//		joyptr=&JOYP_Rj;
//}

void soundset() {
	soundmode++;
	if(soundmode>2)
		soundmode=0;
	Sound_reset_();
}
void centerset() {
	g_scalingx^=1;
}
void timerset() {
	timermask^=1;
}
